drv/sidecar-front-io/src/transceivers.rs

```diff
@@ -0,0 +1,201 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::{Addr, Reg};
+use drv_fpga_api::{FpgaError, FpgaUserDesign, WriteOp};
+use zerocopy::{AsBytes, FromBytes, Unaligned};
+
+pub struct Transceivers {
+    fpgas: [FpgaUserDesign; 2],
+}
+
+impl Transceivers {
+    pub fn new(fpga_task: userlib::TaskId) -> Self {
+        Self {
+            // There are 16 QSFP-DD transceivers connected to each FPGA
+            fpgas: [
+                FpgaUserDesign::new(fpga_task, 0),
+                FpgaUserDesign::new(fpga_task, 1),
+            ],
+        }
+    }
+
+    pub fn get_modules_status(&self) -> Result<[u32; 7], FpgaError> {
+        let mut r: [u32; 7] = [0; 7];
+        let f0: [u8; 14] = self.fpgas[0].read(Addr::QSFP_CTRL_EN_H)?;
+        let f1: [u8; 14] = self.fpgas[1].read(Addr::QSFP_CTRL_EN_H)?;
+
+        for i in 0..7 {
+            r[i] = ((f1[i * 2] as u32) << 24)
+                | ((f1[i * 2 + 1] as u32) << 16)
+                | ((f0[i * 2] as u32) << 8)
+                | (f0[i * 2 + 1] as u32);
+        }
+
+        Ok(r)
+    }
+
+    pub fn masked_op(
+        &self,
+        op: WriteOp,
+        mask: u32,
+        addr: Addr,
+    ) -> Result<(), FpgaError> {
+        let m = mask.as_bytes();
+
+        if (mask & 0xFFFF) > 0 {
```

Comment:
Nit: use `!=` instead of `>` to make it more obvious (otherwise, I have to think about whether this is signed)

Reply:
Good point!